packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/modal/ReactModalHostManager.java

```diff
@@ -136,6 +143,17 @@ public void onShow(DialogInterface dialog) {
                   new ShowEvent(UIManagerHelper.getSurfaceId(reactContext), view.getId()));
             }
           });
+      view.setOnDismissListener(
+          new DialogInterface.OnDismissListener() {
+            @Override
+            public void onDismiss(DialogInterface dialog) {
+              if (mIdentifier != UNSET) {
+                WritableMap args = Arguments.createMap();
+                args.putInt("modalID", mIdentifier);
+                reactContext.emitDeviceEvent("modalDismissed", args);
+              }
+            }
```

Comment:
We don't want to use device events for this. Instead create a DismissEvent class (see ShowEvent/RequestCloseEvent) and dispatch that instead.

Reply:
That doesn't seem right. `visible` is still true when the dismiss event is generated. 

The iOS code notes that this is only used in the old architecture, and you can see a "normal" dismiss event generated in the new arch: https://github.com/facebook/react-native/blob/main/packages/react-native/React/Fabric/Mounting/ComponentViews/Modal/RCTModalHostViewComponentView.mm#L177

Reply:
@javache I tested it on a simple app and I am afraid @bernhardoj is right that we need `emitDeviceEvent` on `Android`. When setting `visible` prop to `false`, we remove the `RCTModalHostView` from React hierarchy, which triggers native dismissal of the component. Therefore, there is no way to have the React component still there when the event is dispatched. It's the same on `iOS` where we end up in here: https://github.com/facebook/react-native/blob/4dbf5dca7ab77deee734dd033a1e5350dfdc6764/packages/react-native/React/Views/RCTModalHostViewManager.m#L95, then https://github.com/facebook/react-native/blob/4dbf5dca7ab77deee734dd033a1e5350dfdc6764/packages/react-native/React/Views/RCTModalManager.m#L38 which calls: https://github.com/facebook/react-native/blob/4dbf5dca7ab77deee734dd033a1e5350dfdc6764/packages/react-native/React/Modules/RCTEventEmitter.m#L66 which sends the event with `RCTDeviceEventEmitter`.

I also checked `iOS` on new architecture, and this event is sent here https://github.com/facebook/react-native/blob/4dbf5dca7ab77deee734dd033a1e5350dfdc6764/packages/react-native/React/Fabric/Mounting/ComponentViews/Modal/RCTModalHostViewComponentView.mm#L177. The problem with it is that since the flow is the same as on old architecture, the `eventEmitter` is already `nullptr` by the time we want to send the event, so it is never received on the JS side. I made a quick fix for it, so @bernhardoj it would be nice if you could add it to your PR or I can make a new one with it: https://gist.github.com/WoLewicki/bbbde2e57733c6bef0e7d83173c703aa.

I checked `Android` on the new architecture and it works with the changes from the PR.

Is there a way we can move it forward?

Reply:
Why is it so with the pressable?

Reply:
I'd stay with the current solution of rendering `null` and sending the device event since this approach works and does not change the logic of the component. Changing it might be an issue for another PR and would probably require more collaboration.

Reply:
@bernhardoj Thanks for investigating this on iOS too. Your fix won't work in the new bridgeless architecture, so we may need to dive a bit deeper here.

Using a native module here is working around the React component model, which does enforce that unmounted components no longer can send events. A more correct solution could be to implement the dismissal event inside `Modal` and send it from `componentDidUpdate` if the modal was previously visible.

Reply:
> A more correct solution could be to implement the dismissal event inside Modal and send it from componentDidUpdate if the modal was previously visible.

So maybe we should try this solution first to have the working code with the minimal changes and then we could think of trying other solutions in some follow-up PRs?